File: LeetCode/0973_K_Closest_Points_to_Origin.cpp
// quick select
// This is known as "quickselecting by a pivot x"
class Solution {
private:
    vector<vector<int>> mypoints;

    int dist(int i) {
        return mypoints[i][0]*mypoints[i][0]+mypoints[i][1]*mypoints[i][1];
    }

    void swap(int i, int j) {
        int temp0 = mypoints[i][0];
        int temp1 = mypoints[i][1];
        mypoints[i][0] = mypoints[j][0];
        mypoints[i][1] = mypoints[j][1];
        mypoints[j][0] = temp0;
        mypoints[j][1] = temp1;
    }

    int partition(int i, int j) {
        int oi=i;
        int pivot=dist(i);
        i++;

        while(true) {
            while(i<j && dist(i) < pivot) i++;
            while(i<=j && dist(j) > pivot) j--;
            if(i>=j) break;
            swap(i,j);
        }
        swap(oi,j);
        return j;
    }

    void sort(int i, int j, int K) {
        if(i>=j) return;
        int oi=i, oj=j;
        int k = i+rand()%(j-i);
        swap(i,k);

        int mid = partition(i,j);
        int leftLength = mid - i + 1;
        if(K < leftLength)
            sort(i, mid-1, K);
        else if(K > leftLength)
            sort(mid+1, j, K-leftLength);
    }



public:
    vector<vector<int>> kClosest(vector<vector<int>>& points, int K) {
        mypoints = points;
        sort(0, points.size()-1, K);
        return vector<vector<int>>(mypoints.begin(), mypoints.begin()+K);
    }

};



// class Solution {
// public:
//     static bool cmp(vector <int> &x, vector <int> &y) {
//         return (x[0] * x[0] + x[1] * x[1]) <= (y[0] * y[0] + y[1] * y[1]);
//     }
//
//     vector<vector<int>> kClosest(vector<vector<int>>& p, int K) {
//         nth_element(p.begin(), p.begin() + K, p.end(), cmp);
//         p.resize(K);
//         return p;
//     }
// };
//
// static int fastio = []() {
//     #define endl '\n'
//     std::ios::sync_with_stdio(false);
//     std::cin.tie(NULL);
//     std::cout.tie(0);
//     return 0;
// }();

File: LeetCode/0973_K_Closest_Points_to_Origin.java
// quick select
// This is known as "quickselecting by a pivot x"
class Solution {

  private int getDist(int[] point) {
      return point[0] * point[0] + point[1] * point[1];
  }

  private void swap(int[][] points, int i, int j) {
      int t0 = points[i][0];
      int t1 = points[i][1];
      points[i][0] = points[j][0];
      points[i][1] = points[j][1];
      points[j][0] = t0;
      points[j][1] = t1;
  }

  private int partition(int[][] points, int l, int r) {
      int tmp = r;
      int pivotDist = getDist(points[r]);
      r -= 1;

      while(l <= r) {
          while (l <= r && getDist(points[l]) < pivotDist)
              l++;
          while (l <= r && getDist(points[r]) >= pivotDist)
              r--;

          if (l <= r) {
              swap(points, l, r);
              l++;
              r--;
          }
      }
      swap(points, tmp, l);
      return l;
  }

  public void sort(int[][] points, int K, int start, int end) {
      if (start >= end) return;

      /* Fancy way to pick pivot
      int k = ThreadLocalRandom.current().nextInt(start, end);
      // swap with the end
      swap(points, k, end);
      */

      // here, we just pick the last as pivot

      int mid = partition(points, start, end);
      int leftLength = mid - start + 1;

      /* Method 1 to switch*/
      if (K < mid) {
          sort(points, K, start, mid - 1);
      }else if (K > mid) {
          sort(points, K, mid + 1, end);
      }


      /* Method 2: Divide and Conquer (Quick Partition)*/
      public int[][] kClosest(int[][] points, int K) {
          int[][] res = new int[K][];
          if (points == null || points.length == 0 || K >= points.length) return res;

          sort(points, K, 0, points.length - 1);

          return Arrays.copyOfRange(points, 0, K);

      }

      /* Method 2 to switch*/
      // if (K < leftLength)
      //     sort(points, K, start, mid - 1);
      // else if (K > leftLength)
      //     sort(points, K, mid + 1, end);

      // else we find right position K == mid

  }
    /* Method 1 : Heap*/
//     public int[][] kClosest(int[][] points, int K) {
//         int[][] res = new int[K][];

//         if (points == null || points.length == 0 || K >= points.length) return res;

//         // max heap
//         PriorityQueue<int[]> pq = new PriorityQueue<int[]>(K + 1, (o1, o2) -> getDist(o2) - getDist(o1));

//         for (int[] point : points) {
//             pq.add(point);
//             if (pq.size() > K) {
//                 pq.poll();
//             }
//         }

//         for (int i = 0; i < K; i++) {
//             res[i] = pq.poll();
//         }

//         return res;

//     }


}
